### PR TITLE
Autogenerate CHANGES.md

### PR DESCRIPTION
**Pull request by franklin-feingold**
PR will close #3 

This solution will collect the merged PR and put it under the unreleased section. It will also update the unreleased section to the new release number after the release has occurred when a new PR has been merged (the next one after the release PR, with that one now being under the unreleased section). The src/CHANGES.md is how it looks ran on this repo in the current state. 

The changelog will only generate and push when there is a merged PR (it will not push on a release PR or regular commits). A release PR will contain one change (changing the Unreleased title to x.x.x [the version number])

The changelog has also been added to mkdocs. Some of the specification documentation was updated to pass the updated .remarkrc file.

To complete the addition - a github token needs to be added to the CircleCI environmental variables named CHANGE_TOKEN. I can do that if this solution looks good before the merge.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -25,3 +25,124 @@ jobs:
...
+           working_directory: ~/build
+           command: |
+             if git log -1 --pretty=%s | grep Merge* ; then
+             if git log -1 --pretty=%b | grep release ; then
+             echo "New release, do nothing"
+             else
+             mv ~/build/tmp/CHANGES.md ~/build/src/CHANGES.md
+             git config credential.helper 'cache --timeout=120'
+             git config user.email "franklin.feingold@gmail.com"
+             git config user.name "franklin-feingold"
```
Comment (chrisfilo) [inline .circleci/config.yml]: Probably better to change it to something like "changelog bot".

File: .remarkrc [hunk 2]
```diff
@@ -1,6 +1,9 @@
 {
   "plugins": [
     "preset-lint-markdown-style-guide",
-    ["lint-no-duplicate-headings", false]
+    ["lint-no-duplicate-headings", false],
+    ["lint-list-item-indent", "tab-size"],
+    ["lint-emphasis-marker", "consistent"],
+    ["lint-maximum-line-length", false]
```
Comment (chrisfilo) [inline .remarkrc]: Why are those changed to the linter necessary for the changelog generation?

File: .circleci/config.yml [hunk 3]
```diff
@@ -25,3 +25,124 @@ jobs:
...
+      - setup_remote_docker:
+           version: 17.11.0-ce
+      - checkout
+      - run:
+           name: Build changelog
+           working_directory: ~/build
+           command: |
+             if git log -1 --pretty=%s | grep Merge* ; then
+             if git log -1 --pretty=%b | grep release ; then
+             echo "New release, do nothing"
```
Comment (chrisfilo) [inline .circleci/config.yml]: Is this check necessary? I wonder if there could be a situation where PR title would include the word "release" (with a different meaning).

File: .circleci/config.yml [hunk 4]
```diff
@@ -25,3 +25,124 @@ jobs:
...
+           npm install remark-cli@5.0.0 remark-lint@6.0.2 remark-preset-lint-recommended@3.0.2 remark-preset-lint-markdown-style-guide@2.1.2
+      - run: # remark the auto generated changes.md
+          name: remark on autogenerated CHANGES.md
+          command: |
+             cd ~/project
+             if git log -1 --pretty=%s | grep Merge* ; then
+             if git log -1 --pretty=%b | grep release ; then
+             echo "New release, do nothing"
+             mkdir ~/project/src/tmp
+             touch ~/project/src/tmp/empty.txt
```
Comment (chrisfilo) [inline .circleci/config.yml]: This seems identical to the "not a pr merge commit" path - logic could be simplified.

**Comment by chrisfilo** [comment]
Great work! Thanks for putting this together!

File: .circleci/config.yml [hunk 5]
```diff
@@ -25,3 +25,124 @@ jobs:
...
+  search_build:
+    jobs:
+      - build
+      - github-changelog-generator
+      - remark:
+          requires:
+               - github-changelog-generator
+      - push:
+          requires:
+               - remark
```
Comment (effigies) [inline .circleci/config.yml]: Do you want to run this on every commit? I think limiting to `branch: master` will make more sense:

```suggestion
               - remark
           filters:
               branches:
                   only: master
```

This should probably apply to `github-changelog-generator` and `remark` as well.

File: .circleci/config.yml [hunk 6]
```diff
@@ -25,3 +25,124 @@ jobs:
...
+             if git log -1 --pretty=%b | grep release ; then
+             echo "New release, do nothing"
+             else
+             mv ~/build/tmp/CHANGES.md ~/build/src/CHANGES.md
+             git config credential.helper 'cache --timeout=120'
+             git config user.email "franklin.feingold@gmail.com"
+             git config user.name "franklin-feingold"
+             git add ~/build/src/CHANGES.md
+             git commit -m "autobuild CHANGES.md"
+             git push https://${CHANGE_TOKEN}@github.com/bids-standard/bids-specification.git master
```
Comment (effigies) [inline .circleci/config.yml]: I might suggest pushing to another branch than `master`, in case you want to modify before pushing to `master`. But if the plan is just to clean up later, then no big deal.

File: .circleci/config.yml [hunk 7]
```diff
@@ -25,3 +25,124 @@ jobs:
...
+             if git log -1 --pretty=%s | grep Merge* ; then
+             if git log -1 --pretty=%b | grep release ; then
+             echo "New release, do nothing"
+             else
+             mv ~/build/tmp/CHANGES.md ~/build/src/CHANGES.md
+             git config credential.helper 'cache --timeout=120'
+             git config user.email "franklin.feingold@gmail.com"
+             git config user.name "franklin-feingold"
+             git add ~/build/src/CHANGES.md
+             git commit -m "autobuild CHANGES.md"
```
Comment (effigies) [inline .circleci/config.yml]: I would prefer a cleaner/more descriptive message. Something like `[DOC] Auto-generate changelog entry for PR #XYZ`.

File: .circleci/config.yml (hunk 3 above)
Comment (franklin-feingold) [reply .circleci/config.yml]: This check is necessary to make sure that the changelog will only generate on merged PR's. It will skip for commits and release PRs. For a release PR title we could have a controlled vocabulary or set of words to denote only a new specification release ("New main specification release"? with the new version followed i.e. New main specification release 1.2.3). I agree there can be a situation where a PR title could include the word "release" in it and mean something else. 

The release tag could work. The git describe --tags outputs the `"latest release tag"-# of commits since release-g"(commit string)"` . This can be parsed down to get the latest release tag, but I noticed that the git tag --list will output whatever tag you provide it, so the test will be true for all cases. I may not see how the test will skip the changelog generation for a new release?

File: .circleci/config.yml (hunk 4 above)
Comment (franklin-feingold) [reply .circleci/config.yml]: Across the two jobs it is a similar path. The jobs have to be separated because of the docker environment differences between the two jobs.

File: .circleci/config.yml (hunk 6 above)
Comment (franklin-feingold) [reply .circleci/config.yml]: The cleaning up happens automatically so it should be all ready for the master branch (with no linter errors). The hope is that this is an automatic process not needing manual intervention. If the automation cannot be realized then pushing to another branch and then be PR'd over to master would work.

File: .remarkrc (hunk 2 above)
Comment (franklin-feingold) [reply .remarkrc]: Those changes are necessary for the auto generated changelog (after it was corrected through remark) to pass the linter style guide

File: .circleci/config.yml (hunk 5 above)
Comment (franklin-feingold) [reply .circleci/config.yml]: I have applied this change in my recent commit

File: .circleci/config.yml (hunk 7 above)
Comment (franklin-feingold) [reply .circleci/config.yml]: I have applied this update to my recent commit

File: .circleci/config.yml (hunk 1 above)
Comment (effigies) [reply .circleci/config.yml]: I think he meant change the name of `user.name`.

File: .circleci/config.yml (hunk 3 above)
Comment (effigies) [reply .circleci/config.yml]: Try in some project with tags (say 0.1.0):

```Bash
for REF in 0.1.0 master; do
    git checkout -q $REF
    DESC=$(git describe --tags)
    if [ "$(git tag --list $DESC)" = "$DESC" ]; then
        echo "We're on a tag: $DESC"
    else
        echo "Not a tag, after all: $DESC"
    fi
done
```

In Pybids, for example:

```
We're on tag: 0.1.0
Not a tag, after all: 0.6.5-229-gbd23eaa
```

This is because `git tag --list 0.6.5-229-gbd23eaa` doesn't match any tag, and produces an empty string.

Comment (effigies) [reply .circleci/config.yml]: Oh, I see. You plan to do release prep PRs. In which case, I would recommend using a template like "REL: <version>" for the PR name (or similar), and then looking for the "REL: " string in the merge commit. Otherwise you'll have to be vigilant about any PRs that have "release" somewhere in the title, which may show up innocently. "REL: " is much less likely to happen by accident.

**Comment by chrisfilo** [comment]
RTM @franklin-feingold ?

**Comment by chrisfilo** [comment]
Many thanks!
